File: Q00201-Q00300/00230-Kth-Smallest-Element-in-a-BST/java00230/Solution2.java
package java00230;

import java.util.ArrayList;
import java.util.List;

/**
 * @Author: alton
 * @Date: Created in 6/12/21 10:33 AM
 * @Description:
 *
 * 230. Kth Smallest Element in a BST #101
 * <p>
 * Given the root of a binary search tree, and an integer k, return the kth (1-indexed) smallest element in the tree.
 * <p>
 * <p>
 * <p>
 * Example 1:
 * <p>
 * <p>
 * Input: root = [3,1,4,null,2], k = 1
 * Output: 1
 * Example 2:
 * <p>
 * <p>
 * Input: root = [5,3,6,2,4,null,null,1], k = 3
 * Output: 3
 * <p>
 * <p>
 * Constraints:
 * <p>
 * The number of nodes in the tree is n.
 * 1 <= k <= n <= 104
 * 0 <= Node.val <= 104
 * <p>
 * <p>
 * Follow up: If the BST is modified often (i.e., we can do insert and delete operations) and you need to find the kth smallest frequently, how would you optimize?
 * <p>
 * Time Complexity: O()
 * Space Complexity: O()
 * /**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 *
 * Runtime: 0 ms, faster than 100.00% of Java online submissions for Kth Smallest Element in a BST.
 * Memory Usage: 39.7 MB, less than 20.41% of Java online submissions for Kth Smallest Element in a BST.
 *
 * Time Complexity: O(N)
 * Space Complexity: O(N)
 *
 * dfs inorder
 */
class Solution2 {

    int kth = -1;
    int res = -1;
    public int kthSmallest(TreeNode root, int k) {
        kth = k;
        inorder(root);

    }

    private void inorder(TreeNode node) {

        if (node == null || kth < 0) {
            return;
        }

        inorder(node.left);

        if (--kth == 0) {
            res = node.val;
        }
        inorder(node.right);

    }
}

File: Q00201-Q00300/00230-Kth-Smallest-Element-in-a-BST/cpp00230/Solution2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 * Runtime: 16 ms, faster than 81.47% of C++ online submissions for Kth Smallest Element in a BST.
 * Memory Usage: 24 MB, less than 85.57% of C++ online submissions for Kth Smallest Element in a BST.
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        kth = k;
        inorder(root);
        return res;
    }

private:
    int res = -1;
    int kth = -1;
    void inorder(TreeNode* node) {

        if (node == nullptr || kth < 0) {
            return;
        }

        inorder(node -> left);

        if (--kth == 0) {
            res = node -> val;
            return;
        }

        inorder(node -> right);

    }
};